hibernate-core/src/main/java/org/hibernate/graph/internal/AttributeNodeImpl.java

```diff
@@ -0,0 +1,327 @@
+
+	@SuppressWarnings("unchecked")
+	private <S extends J> SubGraphImplementor<S> internalMakeSubgraph(Class<S> subType) {
+		verifyMutability();
+
+		final ManagedTypeImplementor<S> managedType = valueGraphTypeAsManaged();
+
+		if ( subType == null ) {
+			subType = managedType.getJavaType();
+		}
+
+		return internalMakeSubgraph( managedType.findSubType( subType ) );
+	}
+
+	@SuppressWarnings({"WeakerAccess", "unchecked"})
+	protected <S extends J> void internalAddSubGraph(Class<S> subType, SubGraphImplementor<S> subGraph) {
+		log.tracef( "Adding sub-graph : ( (%s) %s )", subGraph.getGraphedType().getName(), getAttributeName() );
+
+		if ( subGraphMap == null ) {
+			subGraphMap = new HashMap<>();
+		}
+
+		final SubGraphImplementor<? extends J> previous = subGraphMap.put( subType, (SubGraphImplementor) subGraph );
+		if ( previous != null ) {
+			log.debugf( "Adding sub-graph [%s] over-wrote existing [%]", subGraph, previous );
+		}
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public <S extends J> void addSubGraph(Class<S> subType, SubGraph<S> subGraph) {
+		verifyMutability();
+
+		internalAddSubGraph( subType, (SubGraphImplementor) subGraph );
+	}
+
+	@Override
+	public SubGraphImplementor<J> makeKeySubGraph() {
+		return internalMakeKeySubgraph( (Class<J>) null );
+	}
+
+	@Override
+	public <S extends J> SubGraphImplementor<S> makeKeySubGraph(Class<S> subtype) {
+		return internalMakeKeySubgraph( subtype );
+	}
+
+	@Override
+	public <S extends J> SubGraphImplementor<S> makeKeySubGraph(ManagedTypeImplementor<S> subtype) {
+		return internalMakeKeySubgraph( subtype );
+	}
+
+	private <S extends J> SubGraphImplementor<S> internalMakeKeySubgraph(ManagedTypeImplementor<S> type) {
+
+		log.debugf( "Making key sub-graph : ( (%s) %s )", type.getName(), getAttributeName() );
+
+		final SubGraphImplementor<S> subGraph = type.makeSubGraph();
+		internalAddKeySubGraph( type.getJavaType(), subGraph );
+
+		return subGraph;
+	}
+
+	@SuppressWarnings("unchecked")
+	private <S extends J> SubGraphImplementor<S> internalMakeKeySubgraph(Class<S> type) {
+		verifyMutability();
+
+		final ManagedTypeImplementor<S> managedType = keyGraphTypeAsManaged();
+
+		final ManagedTypeImplementor<S> subType;
+
+		if ( type == null ) {
+			subType = managedType;
+		}
+		else {
+			subType = managedType.findSubType( type );
+		}
+
+		subType.getJavaType();
+
+		return internalMakeKeySubgraph( subType );
+	}
+
+	@SuppressWarnings({"WeakerAccess", "unchecked"})
+	protected <S extends J> void internalAddKeySubGraph(Class<S> subType, SubGraph<S> subGraph) {
+		log.tracef( "Adding key sub-graph : ( (%s) %s )", subType.getName(), getAttributeName() );
+
+		if ( keySubGraphMap == null ) {
+			keySubGraphMap = new HashMap<>();
+		}
+
+		final SubGraphImplementor<? extends J> previous = keySubGraphMap.put( subType, (SubGraphImplementor) subGraph );
+		if ( previous != null ) {
+			log.debugf( "Adding key sub-graph [%s] over-wrote existing [%]", subGraph, previous );
```

Comment:
I don't remember what was in the code I submitted but this should really never happen. It should be handled somewhere but, when adding graph specifications they should be merged. If this is meant to be handled outside then this is an unexpected state probably warranting an exception. If it is not meant to be handled outside then it should be handled here.